src/charm.py

```
@@ -485,45 +479,45 @@ def _open_ports(self) -> None:
-            self._mysql.fix_data_dir(container)
-            self._mysql.initialise_mysqld()
-
-            # Add the pebble layer
-            logger.debug("Adding pebble layer")
-            container.add_layer(MYSQLD_SAFE_SERVICE, self._pebble_layer, combine=False)
-            container.restart(MYSQLD_SAFE_SERVICE)
-
-            logger.debug("Waiting for instance to be ready")
-            self._mysql.wait_until_mysql_connection(check_port=False)
-
-            logger.info("Configuring instance")
-            # Configure all base users and revoke privileges from the root users
-            self._mysql.configure_mysql_users()
-            # Configure instance as a cluster node
-            self._mysql.configure_instance()
+    def _write_mysqld_configuration(self):
+        """Write the mysqld configuration to the file."""
+        memory_limit_bytes = (self.config.profile_limit_memory or 0) * BYTES_1MB
+        new_config_content, _ = self._mysql.render_mysqld_configuration(
+            profile=self.config.profile,
+            memory_limit=memory_limit_bytes,
+        )
+        self._mysql.write_content_to_file(path=MYSQLD_CONFIG_FILE, content=new_config_content)
 
-            if self.has_cos_relation:
-                if container.get_services(MYSQLD_EXPORTER_SERVICE)[
-                    MYSQLD_EXPORTER_SERVICE
-                ].is_running():
-                    # Restart exporter service after configuration
-                    container.restart(MYSQLD_EXPORTER_SERVICE)
-                else:
-                    container.start(MYSQLD_EXPORTER_SERVICE)
-        except (
-            MySQLConfigureInstanceError,
-            MySQLConfigureMySQLUsersError,
-            MySQLInitialiseMySQLDError,
-            MySQLCreateCustomConfigFileError,
-        ) as e:
-            logger.debug("Unable to configure instance: {}".format(e))
-            return False
+    def _configure_instance(self, container) -> None:
```

changed to just raise exception on error, instead of boolean return